tests/tests_lexemes.py

```diff
@@ -1,19 +1,63 @@
-from compiler.lexemes import get_repo_list
+from compiler.lexemes import get_repo_list, parse_for_variable_hierarchies
+from ruamel.yaml.comments import CommentedSeq, CommentedMap
 import unittest
 
-class MyTest(unittest.TestCase):
+class Test_get_repo_list(unittest.TestCase):
     def test(self):
-        input = """
-        lightweight_components:
-  - type: compute_element
-    repository_url: "https://github.com/WLCG-Lightweight-Sites/wlcg_lightweight_site_ce_cream"
-    nodes:
-      - node: lw-site-droplet-0
-        container_count: 1
-      - node: ec2-18-184-37-92.eu-central-1
-        container_count: 2
-    preferred_tech_stack:
-      level_2_configuration: yaim
-      """    
-
-        #self.assertEqual(input, ["https://github.com/WLCG-Lightweight-Sites/wlcg_lightweight_site_ce_cream"])
\ No newline at end of file
+      
+        with open("./tests/resources/complete_config.yaml", "r") as f:
+            self.assertEqual(
+                    get_repo_list(f),
+                    [
+                        "https://github.com/WLCG-Lightweight-Sites/wlcg_lightweight_site_ce_cream",
+                        "https://github.com/WLCG-Lightweight-Sites/wlcg_lightweight_site_wn_pbs"
+                    ])
+
+class Test_parse_for_variable_hierarchies(unittest.TestCase):
```

Comment:
Good work @Boot-Error . Let's try to one-up this test case ;) 
Essentially, we can read from a sample site_level_config_file.yaml which contains different use case scenarios for the __from__ keyword.
```
global_variables:
  - &spark_master_ip_address 188.184.104.25
  - &spark_master_fqdn simple-lc01.cern.ch
  - &spark_worker_ip_address 188.184.30.19
  - &spark_worker_fqdn simple-lc02.cern.ch
  - &spark_submit_ip_address 188.185.84.189
  - &spark_submit_fqdn simple-lc03.cern.ch
  - &spark_level_2_configurator sh
  - &l2_configurator:
    __from__: *spark_level_2_configurator

spark_master_runtime_variables:
  - &master_runtime_var_host
    __from__: *spark_master_fqdn

spark_worker_runtime_variables:
  - &spark_worker_runtime_var_host
    __from__: *spark_worker_fqdn
  - &spark_worker_runtime_var_spark_master
    __from__: *spark_master_fqdn

spark_submit_runtime_variables:
  - &submit_runtime_var_host
    __from__: *spark_submit_fqdn

preferred_tech_stack:
  level_1_configuration: puppet
  level_2_configuration: 
    __from__: *l2_configurator
  container_orchestration: docker-swarm
  container: docker
```
We can use PyYAML to parse this file into a python dict and pass it to the parse_variable_hierarchies function.
We can test the output the the parse_variable_hierarchies function to check if the variables that were assigned values using __from__ get the correct values. Let me know if anything needs clarification.

You can incorporate this once you have drafted your proposal ;) Good Luck.

Reply:
If the config is loaded as a dict, then parse_variable_hierarchies function returns the same dictionary (first if condition).

So it is also important to load the same file using ruamel's yaml loader to test if the CommentSeq and CommentMap conditions are functioning too.

And I'm having trouble with loading the same config using ruamel's loader. It's giving me `ReusedAnchorWarning` to all '&' and fails at `__from__: *l2_configurator`. Here is the full stack trace:

```
/usr/lib/python3.7/site-packages/ruamel.yaml-0.15.81-py3.7-linux-x86_64.egg/ruamel/yaml/composer.py:131: ReusedAnchorWarning:
found duplicate anchor 'spark_master_ip_address'
first occurrence   in "<unicode string>", line 2, column 5:
      - &spark_master_ip_address 188.184 ...
        ^ (line: 2)
second occurrence   in "<unicode string>", line 2, column 5:
      - &spark_master_ip_address 188.184 ...
        ^ (line: 2)
  warnings.warn(ws, ReusedAnchorWarning)
/usr/lib/python3.7/site-packages/ruamel.yaml-0.15.81-py3.7-linux-x86_64.egg/ruamel/yaml/composer.py:131: ReusedAnchorWarning:
found duplicate anchor 'spark_master_fqdn'
first occurrence   in "<unicode string>", line 3, column 5:
      - &spark_master_fqdn simple-lc01.c ...
        ^ (line: 3)
second occurrence   in "<unicode string>", line 3, column 5:
      - &spark_master_fqdn simple-lc01.c ...
        ^ (line: 3)
  warnings.warn(ws, ReusedAnchorWarning)
/usr/lib/python3.7/site-packages/ruamel.yaml-0.15.81-py3.7-linux-x86_64.egg/ruamel/yaml/composer.py:131: ReusedAnchorWarning:
found duplicate anchor 'spark_worker_ip_address'
first occurrence   in "<unicode string>", line 4, column 5:
      - &spark_worker_ip_address 188.184 ...
        ^ (line: 4)
second occurrence   in "<unicode string>", line 4, column 5:
      - &spark_worker_ip_address 188.184 ...
        ^ (line: 4)
  warnings.warn(ws, ReusedAnchorWarning)
/usr/lib/python3.7/site-packages/ruamel.yaml-0.15.81-py3.7-linux-x86_64.egg/ruamel/yaml/composer.py:131: ReusedAnchorWarning:
found duplicate anchor 'spark_worker_fqdn'
first occurrence   in "<unicode string>", line 5, column 5:
      - &spark_worker_fqdn simple-lc02.c ...
        ^ (line: 5)
second occurrence   in "<unicode string>", line 5, column 5:
      - &spark_worker_fqdn simple-lc02.c ...
        ^ (line: 5)
  warnings.warn(ws, ReusedAnchorWarning)
/usr/lib/python3.7/site-packages/ruamel.yaml-0.15.81-py3.7-linux-x86_64.egg/ruamel/yaml/composer.py:131: ReusedAnchorWarning:
found duplicate anchor 'spark_submit_ip_address'
first occurrence   in "<unicode string>", line 6, column 5:
      - &spark_submit_ip_address 188.185 ...
        ^ (line: 6)
second occurrence   in "<unicode string>", line 6, column 5:
      - &spark_submit_ip_address 188.185 ...
        ^ (line: 6)
  warnings.warn(ws, ReusedAnchorWarning)
/usr/lib/python3.7/site-packages/ruamel.yaml-0.15.81-py3.7-linux-x86_64.egg/ruamel/yaml/composer.py:131: ReusedAnchorWarning:
found duplicate anchor 'spark_submit_fqdn'
first occurrence   in "<unicode string>", line 7, column 5:
      - &spark_submit_fqdn simple-lc03.c ...
        ^ (line: 7)
second occurrence   in "<unicode string>", line 7, column 5:
      - &spark_submit_fqdn simple-lc03.c ...
        ^ (line: 7)
  warnings.warn(ws, ReusedAnchorWarning)
/usr/lib/python3.7/site-packages/ruamel.yaml-0.15.81-py3.7-linux-x86_64.egg/ruamel/yaml/composer.py:131: ReusedAnchorWarning:
found duplicate anchor 'spark_level_2_configurator'
first occurrence   in "<unicode string>", line 8, column 5:
      - &spark_level_2_configurator sh
        ^ (line: 8)
second occurrence   in "<unicode string>", line 8, column 5:
      - &spark_level_2_configurator sh
        ^ (line: 8)
  warnings.warn(ws, ReusedAnchorWarning)
/usr/lib/python3.7/site-packages/ruamel.yaml-0.15.81-py3.7-linux-x86_64.egg/ruamel/yaml/composer.py:131: ReusedAnchorWarning:
found duplicate anchor 'l2_configurator:'
first occurrence   in "<unicode string>", line 9, column 5:
      - &l2_configurator:
        ^ (line: 9)
second occurrence   in "<unicode string>", line 9, column 5:
      - &l2_configurator:
        ^ (line: 9)
  warnings.warn(ws, ReusedAnchorWarning)
/usr/lib/python3.7/site-packages/ruamel.yaml-0.15.81-py3.7-linux-x86_64.egg/ruamel/yaml/composer.py:131: ReusedAnchorWarning:
found duplicate anchor 'master_runtime_var_host'
first occurrence   in "<unicode string>", line 13, column 5:
      - &master_runtime_var_host
        ^ (line: 13)
second occurrence   in "<unicode string>", line 13, column 5:
      - &master_runtime_var_host
        ^ (line: 13)
  warnings.warn(ws, ReusedAnchorWarning)
/usr/lib/python3.7/site-packages/ruamel.yaml-0.15.81-py3.7-linux-x86_64.egg/ruamel/yaml/composer.py:131: ReusedAnchorWarning:
found duplicate anchor 'spark_worker_runtime_var_host'
first occurrence   in "<unicode string>", line 17, column 5:
      - &spark_worker_runtime_var_host
        ^ (line: 17)
second occurrence   in "<unicode string>", line 17, column 5:
      - &spark_worker_runtime_var_host
        ^ (line: 17)
  warnings.warn(ws, ReusedAnchorWarning)
/usr/lib/python3.7/site-packages/ruamel.yaml-0.15.81-py3.7-linux-x86_64.egg/ruamel/yaml/composer.py:131: ReusedAnchorWarning:
found duplicate anchor 'spark_worker_runtime_var_spark_master'
first occurrence   in "<unicode string>", line 19, column 5:
      - &spark_worker_runtime_var_spark_ ...
        ^ (line: 19)
second occurrence   in "<unicode string>", line 19, column 5:
      - &spark_worker_runtime_var_spark_ ...
        ^ (line: 19)
  warnings.warn(ws, ReusedAnchorWarning)
/usr/lib/python3.7/site-packages/ruamel.yaml-0.15.81-py3.7-linux-x86_64.egg/ruamel/yaml/composer.py:131: ReusedAnchorWarning:
found duplicate anchor 'submit_runtime_var_host'
first occurrence   in "<unicode string>", line 23, column 5:
      - &submit_runtime_var_host
        ^ (line: 23)
second occurrence   in "<unicode string>", line 23, column 5:
      - &submit_runtime_var_host
        ^ (line: 23)
  warnings.warn(ws, ReusedAnchorWarning)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.7/site-packages/ruamel.yaml-0.15.81-py3.7-linux-x86_64.egg/ruamel/yaml/main.py", line 331, in load
    return constructor.get_single_data()
  File "/usr/lib/python3.7/site-packages/ruamel.yaml-0.15.81-py3.7-linux-x86_64.egg/ruamel/yaml/constructor.py", line 106, in get_single_data
    node = self.composer.get_single_node()
  File "/usr/lib/python3.7/site-packages/ruamel.yaml-0.15.81-py3.7-linux-x86_64.egg/ruamel/yaml/composer.py", line 78, in get_single_node
    document = self.compose_document()
  File "/usr/lib/python3.7/site-packages/ruamel.yaml-0.15.81-py3.7-linux-x86_64.egg/ruamel/yaml/composer.py", line 101, in compose_document
    node = self.compose_node(None, None)
  File "/usr/lib/python3.7/site-packages/ruamel.yaml-0.15.81-py3.7-linux-x86_64.egg/ruamel/yaml/composer.py", line 138, in compose_node
    node = self.compose_mapping_node(anchor)
  File "/usr/lib/python3.7/site-packages/ruamel.yaml-0.15.81-py3.7-linux-x86_64.egg/ruamel/yaml/composer.py", line 217, in compose_mapping_node
    item_value = self.compose_node(node, item_key)
  File "/usr/lib/python3.7/site-packages/ruamel.yaml-0.15.81-py3.7-linux-x86_64.egg/ruamel/yaml/composer.py", line 138, in compose_node
    node = self.compose_mapping_node(anchor)
  File "/usr/lib/python3.7/site-packages/ruamel.yaml-0.15.81-py3.7-linux-x86_64.egg/ruamel/yaml/composer.py", line 217, in compose_mapping_node
    item_value = self.compose_node(node, item_key)
  File "/usr/lib/python3.7/site-packages/ruamel.yaml-0.15.81-py3.7-linux-x86_64.egg/ruamel/yaml/composer.py", line 138, in compose_node
    node = self.compose_mapping_node(anchor)
  File "/usr/lib/python3.7/site-packages/ruamel.yaml-0.15.81-py3.7-linux-x86_64.egg/ruamel/yaml/composer.py", line 217, in compose_mapping_node
    item_value = self.compose_node(node, item_key)
  File "/usr/lib/python3.7/site-packages/ruamel.yaml-0.15.81-py3.7-linux-x86_64.egg/ruamel/yaml/composer.py", line 116, in compose_node
    None, None, 'found undefined alias %r' % utf8(alias), event.start_mark
ruamel.yaml.composer.ComposerError: found undefined alias 'l2_configurator'
  in "<unicode string>", line 29, column 15:
        __from__: *l2_configurator
                  ^ (line: 29)
```